### PR TITLE
Add Pca-bo

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [not yet] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: nevergrad/optimization/optimizerlib.py [hunk 1]
```diff
@@ -1815,6 +1817,104 @@ def __init__(
...
+    def _internal_tell_candidate(self, candidate: p.Parameter, loss: tp.FloatLoss) -> None:
+        # print("BUFFER inside tell = ", self.buffer)
+        print("newX =", self.newX)
+        self.loss.append(loss)
+        print("new_loss =", self.loss)
+        if not self.buffer:
+            if "x_probe" in candidate._meta:
+                # y = candidate._meta["x_probe"]
+                print("Len(newX) = ", len(self.newX))
+                self._pcabo.tell(self.newX, self.loss)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1901]: why record newX while it's already in `candidate._meta["x_probe"])`?

File: nevergrad/optimization/optimizerlib.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
 import numpy as np
 from bayes_opt import UtilityFunction
 from bayes_opt import BayesianOptimization
+from typing import List
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:14]:
```suggestion
```

For consistency, please your `tp.List` instead of `List` (`typing` is already imported as `tp`)

File: nevergrad/optimization/optimizerlib.py [hunk 3]
```diff
@@ -1815,6 +1817,171 @@ def __init__(
...
+        if not self.buffer:
+            self.newX = []
+            self.loss = []
+            candidate = self._pcabo.ask()
+            if not isinstance(candidate, list):
+                candidate = candidate.tolist()
+            self.buffer = candidate
+        x_probe = self.buffer.pop()
+        print("X_PROBE = ", x_probe)
+        print("BUFFER = ", self.buffer)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1877]: please remove the prints when everything is working

File: nevergrad/optimization/optimizerlib.py [hunk 4]
```diff
@@ -1815,6 +1817,171 @@ def __init__(
...
+
+@registry.register
+class BayesOptimBO(base.Optimizer):
+    """BO algorithm from BayesOptim package"""
+
+    no_parallelization = True
+    recast = True
+
+    def __init__(
+        self,
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1912]: what is different in this version? This looks mostly like a copy paste, which must be avoided (it's always a pain to maintain on the long run).

Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1912]: The algorithm comes from bayes-optim package but is BO rather than PCA-BO. Yes, it is a copy of the other one, which calls a different optimizer and tunes the hyperparameters of the GPR model - which is automatically done in the case of PCA-BO.

File: nevergrad/optimization/optimizerlib.py (hunk 3 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1877]: Ok

File: nevergrad/optimization/optimizerlib.py (hunk 2 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py]: Done

File: nevergrad/optimization/optimizerlib.py (hunk 4 above)
Comment (jrapin) [reply nevergrad/optimization/optimizerlib.py:1943]: so the only difference is `self._bo2` instead of `self._pcabo`?
what about adding a parameter `pca: bool` in `ParametrizedPCABO`, and then instantiate `self._alg` with either one of the option?

File: nevergrad/optimization/optimizerlib.py [hunk 5]
```diff
@@ -1815,6 +1816,196 @@ def __init__(
...
+PCABO80 = ParametrizedPCABO(n_components=0.80).set_name("PCABO80", register=True)
+PCABO90 = ParametrizedPCABO(n_components=0.90).set_name("PCABO90", register=True)
+PCABO92 = ParametrizedPCABO(n_components=0.92).set_name("PCABO92", register=True)
+PCABO95 = ParametrizedPCABO(n_components=0.95).set_name("PCABO95", register=True)
+PCABO98 = ParametrizedPCABO(n_components=0.98).set_name("PCABO98", register=True)
+
+# Testing the influence of the DoE size on the performance of PCABO
+PCABO95DoE0 = ParametrizedPCABO(n_components=0.95, prop_doe_factor=0.00).set_name("PCABO95DoE0", register=True)
+PCABO95DoE20 = ParametrizedPCABO(n_components=0.95, prop_doe_factor=0.20).set_name("PCABO95DoE20", register=True)
+PCABO95DoE50 = ParametrizedPCABO(n_components=0.95, prop_doe_factor=0.50).set_name("PCABO95DoE50", register=True)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py]: I'd rather avoid having many different options here, because the registry already contains way too many options, I'd keep the 2 or 3 most useful ones, we can decide which later.

If you need all of them for the experiments, it's easy to create them on the fly.

File: nevergrad/optimization/optimizerlib.py [hunk 6]
```diff
@@ -1815,6 +1816,196 @@ def __init__(
...
+            self.loss = []
+            candidate = self._pcabo.ask()
+            if not isinstance(candidate, list):
+                candidate = candidate.tolist()
+            self.buffer = candidate
+        x_probe = self.buffer.pop()
+        data = np.tan(np.array(x_probe, copy=False))
+        candidate = self.parametrization.spawn_child().set_standardized_data(data)
+        candidate._meta["x_probe"] = x_probe
+        self.newX.append(candidate._meta["x_probe"])
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1868]: I think you should move this to the tell instead, this way it's clearer that it's linked to `self._losses`

File: nevergrad/optimization/optimizerlib.py [hunk 7]
```diff
@@ -1815,6 +1816,196 @@ def __init__(
...
+            verbose=True,
+            n_point=1,  # We start with a sequential procedure, maybe we'll extend in a second moment
+            n_components=n_components,
+            acquisition_optimization={"optimizer": "BFGS"},
+        )
+
+    def _internal_ask_candidate(self) -> p.Parameter:
+        if not self.buffer:
+            self.newX = []
+            self.loss = []
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1859]: let's "hide" it with an underscore, and make it a plural for clarity: `self._losses`

File: nevergrad/optimization/optimizerlib.py (hunk 5 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py]: Done

File: nevergrad/optimization/optimizerlib.py (hunk 7 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1859]: Done

File: nevergrad/optimization/optimizerlib.py (hunk 4 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1943]: Not really.. there is also the tuning of the hyperparameters of the GPR model and some other minor things. But I can merge the two classes.

Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py]: Done

File: nevergrad/optimization/optimizerlib.py [hunk 8]
```diff
@@ -1815,6 +1816,139 @@ def __init__(
...
+        from bayes_optim import RealSpace
+        from bayes_optim.surrogate import GaussianProcess
+
+        lb, ub = 1e-7 - np.pi / 2, np.pi / 2 - 1e-7
+        space = RealSpace([lb, ub]) * self.dimension
+
+        # hyperparameters of the GPR model when pca = False
+        thetaL = 1e-10 * (ub - lb) * np.ones(self.dimension)
+        thetaU = 10 * (ub - lb) * np.ones(self.dimension)
+        model = GaussianProcess(thetaL=thetaL, thetaU=thetaU)  # create the GPR model
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1846]: if this is only useful for pca = False, then move it there

File: nevergrad/optimization/optimizerlib.py [hunk 9]
```diff
@@ -1815,6 +1816,139 @@ def __init__(
...
+    """
+
+    no_parallelization = True
+    recast = True
+
+    # pylint: disable=unused-argument
+    def __init__(
+        self,
+        *,
+        # num_workers: int = 1,
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1931]:
```suggestion
```

File: nevergrad/optimization/optimizerlib.py [hunk 10]
```diff
@@ -1815,6 +1816,139 @@ def __init__(
...
+        init_budget: tp.Optional[int] = None,
+        pca: tp.Optional[bool] = False,
+        n_components: tp.Optional[float] = 0.95,
+        prop_doe_factor: tp.Optional[float] = None,
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1935]: can you explain describe the parameters in the docstrings?
See other examples, in this file there should be many (look for "Parameters")

File: requirements/bench.txt [hunk 11]
```diff
@@ -22,7 +22,10 @@ IOHexperimenter>=0.2.8.7
 cdt>=0.5.23
 koncept>=0.2.2
 tensorflow-estimator>=2.3.0
+bayes-optim>=0.2.4
+scikit-learn>=0.24
 tensorflow>=2.5.0
 image-quality>=1.2.7
 keras>=2.4.3
 compiler_gym>=0.1.8 ; sys_platform == "linux"
+
```
Comment (jrapin) [inline requirements/bench.txt:31]:
```suggestion
```

File: nevergrad/optimization/optimizerlib.py [hunk 12]
```diff
@@ -1815,6 +1816,139 @@ def __init__(
...
+            self._losses = []
+            candidate = self._alg.ask()
+            if not isinstance(candidate, list):
+                candidate = candidate.tolist()
+            self._buffer = candidate
+        x_probe = self._buffer.pop()
+        data = np.tan(np.array(x_probe, copy=False))
+        candidate = self.parametrization.spawn_child().set_standardized_data(data)
+        candidate._meta["x_probe"] = x_probe
+        self._newX.append(candidate._meta["x_probe"])
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1889]: please move this line to `_internal_tell_candidate`. This should add robustness and clarity by filling losses and _newX together.

**Review by jrapin — COMMENTED** [review]
Looks like there are problems with the algorithm (cf CI), do you know where that comes from?

File: nevergrad/optimization/optimizerlib.py (hunk 8 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1846]: Done

File: nevergrad/optimization/optimizerlib.py (hunk 9 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1931]: Done

File: nevergrad/optimization/optimizerlib.py (hunk 10 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1935]: Done

File: requirements/bench.txt (hunk 11 above)
Comment (elenaraponi) [reply requirements/bench.txt:31]: Done

File: nevergrad/optimization/optimizerlib.py (hunk 12 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1894]: candidate is returned by `_internal_ask_candidate`, don't we need to apply `tan` and `set_standardized_data` before evaluation of the candidate?

File: nevergrad/optimization/optimizerlib.py (hunk 6 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py]: Answered below

File: nevergrad/optimization/optimizerlib.py (hunk 12 above)
Comment (jrapin) [reply nevergrad/optimization/optimizerlib.py:1894]: I don't understand how your question is related to this line (`self._newX.append(candidate._meta["x_probe"])`)
This will do just the same in tell, storing the value of `x_probe` as it is, the way you normalized it.

Still, the question is interesting, I think there is something fishy here. What is the domain of the "ask" in bayes_optim? if you apply tan on it, it should be -pi/2 to pi/2 to map it to R, I don't think it is

Comment (jrapin) [reply nevergrad/optimization/optimizerlib.py:1894]: For the transform, let's use this:
https://github.com/facebookresearch/nevergrad/blob/fdb364181e9bd3a2d68ad75e1a7ae0605e7ed8f4/nevergrad/parametrization/transforms.py#L218-L244

If your domain is [0, 1], then create `self._transform = ArctanBound(0, 1)`, and `self._transform.backward(data)` will map from [0, 1] to R (and `self._transform.backward` will do the other way round)

File: nevergrad/optimization/optimizerlib.py [hunk 13]
```diff
@@ -1815,6 +1816,150 @@ def __init__(
...
+        parametrization: IntOrParameter,
+        budget: tp.Optional[int] = None,
+        num_workers: int = 1,
+        *,
+        init_budget: tp.Optional[int] = None,
+        pca: tp.Optional[bool] = False,
+        n_components: tp.Optional[float] = 0.95,
+        prop_doe_factor: tp.Optional[float] = None,
+    ) -> None:
+        super().__init__(parametrization, budget=budget, num_workers=num_workers)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1831]:
```suggestion
            config: tp.Optional["ParametrizedBayesOptimBO"] = None,
    ) -> None:
        self._config = ParametrizedBayesOptimBO() if config is None else 
        super().__init__(parametrization, budget=budget, num_workers=num_workers)
```

File: nevergrad/optimization/optimizerlib.py [hunk 14]
```diff
@@ -1815,6 +1816,150 @@ def __init__(
...
+    # pylint: disable=unused-argument
+    def __init__(
+        self,
+        *,
+        init_budget: tp.Optional[int] = None,
+        pca: tp.Optional[bool] = False,
+        n_components: tp.Optional[float] = 0.95,
+        prop_doe_factor: tp.Optional[float] = None,
+    ) -> None:
+        super().__init__(_BayesOptim, locals())
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1948]:
```suggestion
        super().__init__(_BayesOptim, locals(), as_config=True)
```

File: nevergrad/optimization/optimizerlib.py [hunk 15]
```diff
@@ -1815,6 +1816,150 @@ def __init__(
...
+        lb, ub = 1e-7 - np.pi / 2, np.pi / 2 - 1e-7
+        space = RealSpace([lb, ub]) * self.dimension
+
+        self._buffer: tp.List[float] = []
+        self._newX: tp.List[float] = []
+        self._losses: tp.List[float] = []
+
+        # Setting DoE size as a percentage of the total budget if prop_doe_factor is not None
+        if prop_doe_factor and budget:
+            init_budget = round(prop_doe_factor * budget)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1845]:
```suggestion
        init_budget = self._config.init_budget
        if elf._config.prop_doe_factor and budget:
            init_budget = round(elf._config.prop_doe_factor * budget)
```

File: nevergrad/optimization/optimizerlib.py [hunk 16]
```diff
@@ -1815,6 +1816,154 @@ def __init__(
...
+                model=model,
+                DoE_size=init_budget if init_budget is not None else 5,
+                max_FEs=budget,
+                verbose=True,
+            )
+
+    def _internal_ask_candidate(self) -> p.Parameter:
+        if not self._buffer:
+            self._newX = []
+            self._losses = []
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1880]: move those 2 lines right after `self._alg.tell`, it will be more straighforward

File: nevergrad/optimization/optimizerlib.py [hunk 17]
```diff
@@ -1815,6 +1816,154 @@ def __init__(
...
+PCABO = ParametrizedBayesOptim(pca=True).set_name("PCABO", register=True)
+BayesOptimBO = ParametrizedBayesOptim().set_name("BayesOptimBO", register=True)
+
+# Testing the influence of n_components on the performance of PCABO
+PCABO80 = ParametrizedBayesOptim(pca=True, n_components=0.80).set_name("PCABO80", register=True)
+
+# Testing the influence of the DoE size on the performance of PCABO
+PCABO95DoE20 = ParametrizedBayesOptim(pca=True, n_components=0.95, prop_doe_factor=0.20).set_name(
+    "PCABO95DoE20", register=True
+)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1964]: If those are just for testing, either define them in the experiment directly, or in experimentalvariants.py

File: nevergrad/optimization/optimizerlib.py [hunk 18]
```diff
@@ -1815,6 +1816,154 @@ def __init__(
...
+            else:
+                data = candidate.get_standardized_data(reference=self.parametrization)
+                # Tell not asked:
+                self._alg.tell(np.arctan(data), loss)
+
+    def _internal_tell_not_asked(self, candidate: p.Parameter, loss: tp.FloatLoss) -> None:
+        raise errors.TellNotAskedNotSupportedError
+
+
+class ParametrizedBayesOptim(base.ConfiguredOptimizer):
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1906]: let's call this just `BayesOptim`.
Please also import it in optimization/families.py I think this way it will be automatically documented

Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1909]: Why woulf you remove "Parametrized"? We can actually choose a couple of parameters for PCABO: 
```
self.n_components = n_components
self.prop_doe_factor = prop_doe_factor
```

File: nevergrad/optimization/optimizerlib.py (hunk 17 above)
Comment (elenaraponi) [reply nevergrad/optimization/optimizerlib.py:1967]: Moved them to experimentalvariants.py

File: nevergrad/optimization/optimizerlib.py [hunk 19]
```diff
@@ -1815,6 +1816,149 @@ def __init__(
...
+
+    def _internal_ask_candidate(self) -> p.Parameter:
+        if not self._buffer:
+            candidate = self._alg.ask()
+            if not isinstance(candidate, list):
+                candidate = candidate.tolist()
+            self._buffer = candidate
+        x_probe = self._buffer.pop()
+        data = self._transform.backward(np.array(x_probe, copy=False))
+        print("data = ", data)
```
Comment (jrapin) [inline nevergrad/optimization/optimizerlib.py:1887]:
```suggestion
```

**Review by jrapin — APPROVED** [review]
